src/tools/cgclassify.c

```diff
@@ -198,7 +231,449 @@ int main(int argc, char *argv[])
 		/* if any group change fails */
 		if (ret)
 			exit_code = 1;
+
+		/* skip replacing of idle_thread in systemd slice */
+		if (!replace_idle)
+			continue;
+
+		/* systemd idle_thread is already replaced */
+		if (skip_replace_idle)
+			continue;
+
+		scope_pid = find_scope_pid(pid, 0);
+		if (scope_pid == -1) {
+			exit_code = 1;
+			goto err;
+		}
+		skip_replace_idle = 1;
+
+		ret = kill(scope_pid, SIGTERM);
+		if (ret) {
+			err("Failed to kill pid %u:%s\n", scope_pid, strerror(errno));
+			exit_code = 1;
+			goto err;
+		}
 	}
 
 	return exit_code;
+
+err:
+	exit_code = rollback_pid_cgroups(pid);
+	return exit_code;
+}
+
+static pid_t search_systemd_idle_thread_task(pid_t pids[], size_t size)
+{
+	char task_cmd[FILENAME_MAX];
+	char buffer[FILENAME_MAX];
+	FILE *pid_cmd_fp = NULL;
+	int scope_pid = -1;
+	int i;
+
+	for (i = 0; i < size; i++) {
+		snprintf(buffer, FILENAME_MAX, "/proc/%u/cmdline", pids[i]);
+		pid_cmd_fp = fopen(buffer, "re");
+		/* task might have exited */
+		if (!pid_cmd_fp)
+			continue;
+
+		/* task might have exited, so consider only successful reads. */
+		if (fgets(task_cmd, FILENAME_MAX, pid_cmd_fp)) {
+			if (!strcmp(task_cmd, SYSTEMD_IDLE_THREAD)) {
+				scope_pid = pids[i];
+				fclose(pid_cmd_fp);
+				break;
+			}
+		}
+		fclose(pid_cmd_fp);
+	}
+
+	return scope_pid;
+}
+
+static pid_t find_scope_pid(pid_t pid, int capture)
+{
+	pid_t _scope_pid = -1, scope_pid = -1;
+	char ctrl_name[CONTROL_NAMELEN_MAX];
+	char cgroup_name[FILENAME_MAX];
+	char scope_name[FILENAME_MAX];
+	int found_systemd_cgrp = 0;
+	int found_unified_cgrp = 0;
+	char buffer[FILENAME_MAX];
+	FILE *pid_proc_fp = NULL;
+	char *_ctrl_name = NULL;
+	int idx, ret, size = 0;
+	pid_t *pids;
+	int i=0;
+
+	/*
+	 * Let's parse the cgroup of the pid, to check if its in one or
+	 * more .scopes.
+	 */
+	snprintf(buffer, FILENAME_MAX, "/proc/%u/cgroup", pid);
+	pid_proc_fp = fopen(buffer, "re");
+	if (!pid_proc_fp) {
+		err("Failed to open: %s\n", buffer);
+		return -1;
+	}
+
+	while (fgets(buffer, FILENAME_MAX, pid_proc_fp)) {
+		/* read according to the cgroup mode */
+		if (strstr(buffer, "::")) {
+			snprintf(ctrl_name, CONTROL_NAMELEN_MAX, "unified");
+			ret = sscanf(buffer, "%d::%4096s\n", &idx, cgroup_name);
+		} else{
+			ret = sscanf(buffer, "%d:%[^:]:%4096s\n", &idx, ctrl_name, cgroup_name);
+		}
+
+		if (ret != 2 && ret != 3) {
+			err("Unrecognized cgroup file format: %s\n", buffer);
+			goto out;
+		}
+
+		/* cgroup v1 might have shared mount points cpu,cpuacct */
+		_ctrl_name = strchr(ctrl_name, ',');
+		if (_ctrl_name) {
+			size = strlen(ctrl_name) - strlen(_ctrl_name);
+			ctrl_name[size] = '\0';
+		}
+
+		/*
+		 * capture is true, while the pid's controller and cgroups
+		 * are populated for rollback case.
+		 */
+		if (capture) {
```

Comment:
We should probably add a sanity check here:

```
if (i > MAX_MNT_ELEMENTS) {
    err("something or other\n");
    // I'm not sure if we should set the scope_pid back to -1 or do other things
    goto out;
}
```

Reply:
Sure, this sounds like a good idea to add a safety check.